rpyc/lib/win_unix_socket.py

```diff
@@ -0,0 +1,535 @@
+            self._timeout = seconds
+            is_async = ctypes.wintypes.ULONG(1)
+        ioctl(self._sock_fd, FIONBIO, ctypes.byref(is_async))
+
+    def setblocking(self, flag):
+        if flag:
+            self.settimeout(None)
+        else:
+            self.settimeout(0)
+
+    def bind(self, path):
+        """
+        Bind to a path.
+        """
+        path = bytes(path)
+        nbytes = len(path) + 1
+        if nbytes > MAX_PATH_LENGTH:
+            raise socket.error(ERROR_BAD_PATHNAME,
+                    "Socket path name too long ({} bytes, max {})".format(nbytes, MAX_PATH_LENGTH)
+            )
+        addr = sockaddr_un()
+        addr.sun_family = socket.AF_UNIX
+        # copy the name including the null terminator (cpython has null-terminated strings under the hood)
+        ctypes.memmove(addr.sun_path, path, nbytes)
+        if bind(self._sock_fd, ctypes.byref(addr), ctypes.sizeof(addr)) != 0:
+            raise_socket(errno_fn())
+
+    def listen(self, n):
+        """
+        Start to listen on the named pipe.
+        """
+        if listen(self._sock_fd, n) != 0:
+            raise_socket(errno_fn())
+
+    def accept(self):
+        """
+        Block until an incoming connection arrives, or until the timeout elapses.
+
+        Returns a pair with the socket and the peer address.
+        Note: the peer is normally anonymous; a blank string is normal.
+        """
+        addr = sockaddr_un()
+        n = ctypes.c_int(ctypes.sizeof(addr))
+        nonlocal_fd = [None]
+        def do_accept():
+            """
+            Do the accept, return 0 and write to the_fd[0] on success.
+
+            call_socket_fn will handle interrupts, timeouts, error handling.
+            """
+            s = accept(self._sock_fd, ctypes.byref(addr), ctypes.byref(n))
+            if int(s) >= 0:
+                nonlocal_fd[0] = s
+                return 1
+            return 0
+        call_socket_fn(self, 0, do_accept)
+        return (WindowsUnixSocket(fd = nonlocal_fd[0]), addr.path)
+
+    def shutdown(self, how):
+        """
+        Send any data that still needs to be sent.
+
+        This is blocking no matter what, no timeout.
+        """
+        shutdown(self._sock_fd, how)
+
+    def close(self):
+        """
+        Close the socket.
+        """
+        close(self._sock_fd)
+        self._sock_fd = -1
+
+    def connect(self, path):
+        """
+        Connect to a server whose listener pipe is the given path.
+
+        Returns None. After this call, this socket will be connected to the server
+        unless an exception was raised.
+        """
+        path = bytes(path)
+        # copy the name including the null terminator (cpython has null-terminated strings under the hood)
+        nbytes = len(path) + 1
+        if nbytes > MAX_PATH_LENGTH:
+            raise socket.error(ERROR_BAD_PATHNAME,
+                    "Socket path name too long ({} bytes, max {})".format(nbytes, MAX_PATH_LENGTH)
```

Comment:
I remember you refactored this to avoid the duplication; are we working from the correct branch?